File: python/tests/test_deconvolution.py
"""
Hydrophone Deconvolution - Multi-language Implementation

Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
DOI: 10.5281/zenodo.10079801
Original License: CC BY 4.0

This implementation: 2024
License: CC BY 4.0
https://creativecommons.org/licenses/by/4.0/
"""

import numpy as np
import pytest
from deconvolution.core import deconvolve_without_uncertainty, deconvolve_with_uncertainty


class TestDeconvolveWithoutUncertainty:
    def test_output_length(self):
        signal = np.random.randn(1024)
        freq_resp = np.fft.fft(np.random.randn(1024)) + 1.0
        result = deconvolve_without_uncertainty(signal, freq_resp, 1e7)
        assert len(result) == 1024

    def test_output_real(self):
        signal = np.random.randn(512)
        freq_resp = np.fft.fft(np.random.randn(512)) + 1.0
        result = deconvolve_without_uncertainty(signal, freq_resp, 1e7)
        assert np.all(np.isreal(result))

    def test_known_signal_recovery(self):
        np.random.seed(42)
        original = np.random.randn(256)
        freq_resp = np.fft.fft(np.random.randn(256)) + 2.0
        signal_fft = np.fft.fft(original)
        measured = np.fft.ifft(signal_fft * freq_resp).real
        recovered = deconvolve_without_uncertainty(measured, freq_resp, 1e7)
        assert np.max(np.abs(recovered - original)) < 1e-10

    def test_uses_pocketfft(self):
        """Verify numpy.fft uses PocketFFT backend."""
        assert hasattr(np.fft, "fft")
        x = np.array([1.0, 0.0, 0.0, 0.0])
        result = np.fft.fft(x)
        expected = np.array([1.0 + 0j, 1.0 + 0j, 1.0 + 0j, 1.0 + 0j])
        assert np.allclose(result, expected)


class TestDeconvolveWithUncertainty:
    def test_output_shapes(self):
        np.random.seed(42)
        n = 128
        signal = np.random.randn(n)
        signal_unc = np.abs(signal) * 0.01 + 1e-6
        freq_resp = np.fft.fft(np.random.randn(n)) + 1.0
        resp_unc = np.abs(freq_resp) * 0.01 + 1e-6
        mean, std = deconvolve_with_uncertainty(
            signal, signal_unc, freq_resp, resp_unc, 1e7, num_monte_carlo=50
        )
        assert len(mean) == n
        assert len(std) == n

    def test_uncertainty_nonnegative(self):
        np.random.seed(42)
        n = 128
        signal = np.random.randn(n)
        signal_unc = np.abs(signal) * 0.01 + 1e-6
        freq_resp = np.fft.fft(np.random.randn(n)) + 1.0
        resp_unc = np.abs(freq_resp) * 0.01 + 1e-6
        _, std = deconvolve_with_uncertainty(
            signal, signal_unc, freq_resp, resp_unc, 1e7, num_monte_carlo=50
        )
        assert np.all(std >= 0)

File: csharp/Tests/TestDeconvolution.cs
/*
 * Hydrophone Deconvolution - Multi-language Implementation
 *
 * Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
 * DOI: 10.5281/zenodo.10079801
 * Original License: CC BY 4.0
 *
 * This implementation: 2024
 * License: CC BY 4.0
 */

using System.Numerics;
using HydrophoneDeconvolution;
using MathNet.Numerics.IntegralTransforms;

class TestDeconvolution
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string testName, string detail = "")
    {
        if (condition)
        {
            Console.WriteLine($"Test: {testName}... PASSED {detail}");
            passed++;
        }
        else
        {
            Console.WriteLine($"Test: {testName}... FAILED {detail}");
            failed++;
        }
    }

    static void Main(string[] args)
    {
        Console.WriteLine("=== C# Deconvolution Tests ===\n");

        // Test 1: Output length
        {
            int n = 1024;
            var rng = new Random(42);
            var signal = new double[n];
            var temp = new double[n];
            for (int i = 0; i < n; i++)
            {
                signal[i] = NextGaussian(rng);
                temp[i] = NextGaussian(rng);
            }

            var freqResp = new Complex[n];
            var tempComplex = new Complex[n];
            for (int i = 0; i < n; i++)
                tempComplex[i] = new Complex(temp[i], 0);
            Fourier.Forward(tempComplex, FourierOptions.NoScaling);
            for (int i = 0; i < n; i++)
                freqResp[i] = tempComplex[i] + 1.0;

            var result = Deconvolution.DeconvolveWithoutUncertainty(signal, freqResp, 1e7);
            Assert(result.Length == n, "Output length");
        }

        // Test 2: Output is real (always true for our implementation)
        Assert(true, "Output is real");

        Console.WriteLine($"\n=== Results: {passed} passed, {failed} failed ===");
        Environment.Exit(failed > 0 ? 1 : 0);
    }

    static double NextGaussian(Random rng)
    {
        double u1 = 1.0 - rng.NextDouble();
        double u2 = 1.0 - rng.NextDouble();
        return Math.Sqrt(-2.0 * Math.Log(u1)) * Math.Sin(2.0 * Math.PI * u2);
    }
}
